solver/src/paraswap_solver/api.rs

```diff
@@ -0,0 +1,420 @@
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+
+use model::u256_decimal;
+use web3::types::Bytes;
+
+const BASE_URL: &str = "https://apiv4.paraswap.io";
+
+/// Mockable implementation of the API for unit test
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+pub trait ParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse>;
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse>;
+}
+
+struct DefaultParaswapApi {
+    client: Client,
+}
+
+#[async_trait::async_trait]
+impl ParaswapApi for DefaultParaswapApi {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse> {
+        reqwest::get(query.into_url())
+            .await
+            .context("PriceQuery failed")?
+            .json::<Value>()
+            .await
+            .context("PriceQuery result parsing failed")
+            .and_then(PriceResponse::try_from)
+    }
+    async fn transaction(
+        &self,
+        query: TransactionBuilderQuery,
+    ) -> Result<TransactionBuilderResponse> {
+        query
+            .into_request(&self.client)
+            .send()
+            .await
+            .context("TransactionBuilderQuery failed")?
+            .json()
+            .await
+            .context("TransactionBuilderQuery result parsing failed")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum Side {
+    Buy,
+    Sell,
+}
+
+/// Paraswap price quote query parameters.
+#[derive(Clone, Debug)]
+pub struct PriceQuery {
+    /// source token address
+    pub from: H160,
+    /// destination token address
+    pub to: H160,
+    /// decimals of from token (according to API needed  to trade any token)
+    pub from_decimals: usize,
+    /// decimals of to token (according to API needed to trade any token)
+    pub to_decimals: usize,
+    /// amount of source token (in the smallest denomination, e.g. for ETH - 10**18)
+    pub amount: U256,
+    // Type of order
+    pub side: Side,
+}
+
+impl PriceQuery {
+    pub fn into_url(self) -> Url {
+        let mut url = Url::parse(BASE_URL)
+            .expect("invalid base url")
+            .join("v2/prices")
+            .expect("unexpectedly invalid URL segment");
+
+        let side = match self.side {
+            Side::Buy => "BUY",
+            Side::Sell => "SELL",
+        };
+
+        url.query_pairs_mut()
+            .append_pair("from", &format!("{:#x}", self.from))
+            .append_pair("to", &format!("{:#x}", self.to))
+            .append_pair("amount", &self.amount.to_string())
+            .append_pair("side", side)
+            .append_pair("network", "1");
+        url
+    }
+}
+
+/// A Paraswap API price response.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PriceResponse {
+    pub price_route: Map<String, Value>,
+    pub src_amount: U256,
+    pub dest_amount: U256,
+}
+
+impl TryFrom<Value> for PriceResponse {
```

Comment:
Why not implement `Deserialize` instead of `TryFrom<Value>`? You should even be able to derive it with `#[serde(with = "u256_decimal")]`.

If a type implements `Deserialize`, then you can also use [`serde_json::from_value()`](https://docs.rs/serde_json/1.0.64/serde_json/fn.from_value.html) to convert a `serde_json::Value` to that type.

Reply:
I originally tried implementing my own Deserializer. I thought I could just `deserialize_str` like we already do it [here](https://github.com/gnosis/oba-services/blob/16e30b88f9d89198d4ad607a6f5efec5d61b0365/model/src/h160_hexadecimal.rs#L39-L40) but it complained that it would have to use `deserialize_map`. Then, for a [MapAccess](https://docs.serde.rs/serde/de/trait.MapAccess.html) visitor the code looked so complicated and indirect that I gave up and went for this naive approach. I'd be happy to take a look together with you (or take a pointer to where this has already be done). 

I gotta say serde::Deserialize is definitely not my favorite trait.

Reply:
Implementing `Deserialize` by hand is hard.

> I thought I could just deserialize_str [...] but it complained that it would have to use deserialize_map

This would indicate that your type expects the next value from the deserializer to be a string (which is true for `H160` for example because its represented as a string hex value in JSON), but since this is an object that is being deserialized, the nex value from the deserializer should be a map.

Reply:
Will add the following comment (I though I already mentioned it somewhere, but cannot find it)

> The reason we do not `#[derive(Deserialize)]` is that the structure of `price_route` is opaque and may change over time. In order to get the transaction data from the API, it expects `price_route` to be echoed back in the exact form as it was sent in the price estimation response. We therefore leave it as an opaque mapping and only extract the effective in and out amounts from the struct (which we hope will remain in the same place). We don't implement `Deserialize` manually as writing a custom MapVisitor leads to convoluted code.

Reply:
I would still advocate for deriving the deserialization code. You can just keep a field with the "raw" value for serialization and deserialization:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1da79c937b3e0bc135b878be12a06d7c